Removal of  TRestRawZeroSuppressionProcess, moving inside TRestRawToDetectorSignalProcess

juanangp opened the pull request:
![juanangp](https://badgen.net/badge/PR%20submitted%20by%3A/juanangp/blue) ![Large: 292](https://badgen.net/badge/PR%20Size/Large%3A%20292/red) [![](https://gitlab.cern.ch/rest-for-physics/framework/badges/alphaAna/pipeline.svg)](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/alphaAna)<!-- PR-BADGE: PLEASE DO NOT REMOVE THIS COMMENT -->

- TRestRawSignalToSignalProcess renamed to TRestRawToDetectorSignalProcess.
- Removal of TRestRawZeroSuppressionProcess, moving it inside TRestRawToDetectorSignalProcess.
- TRestRawToDetectorSignalProcess. Cleaning up and added threshold parameter.
- TRestDetectorHitsToTrackProcess. Documented, and added official headers.

juanangp committed:
Adding new class TRestDetectorRawSignalToHitsProcess which translates TRestRawSignal into TRestDetectorHitsEvent

jgalan commented on src/TRestDetectorRawSignalToHitsProcess.cxx:
```diff
@@ -0,0 +1,324 @@
+///
+/// History of developments:
+///
+///
```
Missing

```
/// 2022-January: First concept and implementation of this process
/// \author     Juanan Garcia
```

jgalan commented:
Hi, What I don't like of this process is that now there are two paths (not two different methods) to go from a `TRestRawSignalEvent` to a `TRestDetectorHitsEvent`.

This will require twice maintenance, and derive in twice the amount of bugs. If we focus on maintaining common processes we will gain robustness.

A possibility would be that `TRestDetectorSignalEvent` becomes obsolete, however the advantage of having both, TRestDetectorSignal and TRestDetectorHit is that we keep the `readout` business contained in a single library at rest-for-physics/detectorlib, which for organisation and structure I believe it gets its benefits.

jgalan commented:
On the other hand, there is anything at this process that cannot be done with the combination of `TRestRawSignalZeroSuppressionProcess` and `TRestDetectorSignalToHitsProcess`?

juanangp commented:
> 

I really don't like TRestRawSignalZeroSuppressionProcess, I think it should just return a data container similar TRestRawSignalEvent but after applying zero suppression e.g. mimic the zero suppression of the electronics, afterwards you can perform the data analysis that you consider...

juanangp commented:
> Hi, What I don't like of this process is that now there are two paths (not two different methods) to go from a `TRestRawSignalEvent` to a `TRestDetectorHitsEvent`.
> 
> This will require twice maintenance, and derive in twice the amount of bugs. If we focus on maintaining common processes we will gain robustness.
> 
> A possibility would be that `TRestDetectorSignalEvent` becomes obsolete, however the advantage of having both, TRestDetectorSignal and TRestDetectorHit is that we keep the `readout` business contained in a single library at rest-for-physics/detectorlib, which for organisation and structure I believe it gets its benefits.

I agree with you, I think TRestDetectorHits should became obsolete and we should definetively remove code duplication.
The problem is that I don't want to break anything because I assume mostly everybody uses TRestDetectorHits, so I will try to find an intermediate solution.

jgalan commented:
> I really don't like TRestRawSignalZeroSuppressionProcess, I think it should just return a data container similar TRestRawSignalEvent but after applying zero suppression e.g. mimic the zero suppression of the electronics, afterwards you can perform the data analysis that you consider...

Ok, that would be possible, that this process just brings to zero the values of the bins that are not considered signal, and just return a TRestRawSignalEvent as an internal process. Then, TRestRawSignalToDetectorSignalProcess could just remove those zero-ed points or do something else to transform to a physical time. I see nothing wrong with that approach. Just that it needs to be done.

In that case `TRestRawSignalZeroSuppressionProcess` should be moved to rest-for-physics/rawlib.

nkx111 commented:
> I agree with you, I think TRestDetectorHits should became obsolete and we should definetively remove code duplication.

which should become obsolete? TRestDetectorHits or TRestDetectorSignalEvent?

jgalan commented:
> I agree with you, I think TRestDetectorHits should became obsolete 

This is not what I said. Of course, it would be an option to make `TRestDetectorSignal` or `TRestDetectorHits` obsolete, but of course I do not agree it is the best option. We have few classes properly documented, `TRestDetectorSignalToHitsProcess` is one of them. https://sultan.unizar.es/rest/classTRestDetectorSignalToHitsProcess.html

I dont understand the meaning of getting obsolete here. The only option is to upgrade, why obsolete?

jgalan commented:
Keeping `TRestDetectorSignal` and `TRestDetectorHits` philosophy to keep readout business contained in a single entity/library is something we should preserve. Of course, I am open to evolution and upgrade of this scheme.

jgalan commented:
I believe what `TRestDetectorRawSignalToHitsProcess` is doing already exists, so it would be better to understand why what it exists does not fit you.

See also: https://sultan.unizar.es/rest/classTRestRawSignalToSignalProcess.html

A combination of `TRestRawSignalToSignalProcess` and `TRestDetectorSignalToHitsProcess` should do the job. And there will be occasions where someone will be willing to get a `TRestDetectorSignalEvent`. If is clear that `TRestDetectorSignalEvent` is not used at all, then we could think of removing it. But in my opinion it is not the case, because it helps to keep readout business contained within `detectorlib`. We should work to move readout related staff into `detectorlib` and outside `connectorslib`.

nkx111 commented:
> If is clear that TRestDetectorSignalEvent is not used at all

For me `TRestDetectorSignalEvent` is very useful. It helps to reduce the storage of REST files. We usually run some "rough analysis" after data taking. The process chain ends up in the `TRestDetectorSignalEvent` stage. Further analysis can be based on that "roughly analyzed file". Storing output event in `TRestDetectorSignalEvent` type reduces 3~5 times of disk space, compared to `TRestDetectorRawSignalEvent`.  

So in my opinion, neither `TRestDetectorSignalEvent` nor `TRestDetectorHitsEvent` shall be obsolete.

juanangp committed:
Removing TRestDetectorRawSignalToHitsProcess, implementing integral window method in TRestDetectorSignalToHitsProcess

juanangp committed:
Moving TRestRawZeroSuppresionProcess from connectors to raw library

juanangp committed:
Removal of TRestRawZeroSuppresionProcess from CMakelist

juanangp commented:
Ok, now I have implemented the new method in TRestDetectorSignalToHitsProcess rest-for-physics/detectorlib#33

Then I have moved TRestRawZeroSuppressionProcess into rawlib rest-for-physics/rawlib#46

jgalan committed:
TRestRawSignalToSignalProcess renamed to TRestRawToDetectorSignalProcess

jgalan committed:
TRestRawToDetectorSignalProcess. Cleaning up and added threshold parameter. Class version increased

jgalan committed:
TRestDetectorHitsToTrackProcess. Documented, added official headers, cleaning up

jgalan committed:
TRestDetectorHitsToTrackProcess. Fixing compilation issue

juanangp commented on src/TRestRawToDetectorSignalProcess.cxx:
```diff
@@ -167,20 +124,11 @@ TRestEvent* TRestRawSignalToSignalProcess::ProcessEvent(TRestEvent* evInput) {
         sgnl.SetID(rawSgnl->GetID());
         for (int p = 0; p < rawSgnl->GetNumberOfPoints(); p++)
-            sgnl.NewPoint(fTriggerStarts + fSampling * p, fGain * rawSgnl->GetData(p));
+            if (rawSgnl->GetData(p) > fThreshold)
```
This commit breaks the novel "intwindow" analysis, I would suggest to set to zero in case is below threshold

jgalan replied:
No, why to set to zero, we want to save storage space. It is probably something that needs to be fixed at the `intwindow` method. You probably need to define an integration window.

See for example how it is done at:

https://sultan.unizar.es/rest/classTRestDetectorSignalToRawSignalProcess.html

juanangp committed:
Adding zero value when data is below threshold, it was creating issues with new intwindow method

juanangp committed:
Adding zero suppression method to TRestRawToDetectorSignalProcess

juanangp committed:
Merge branch 'master' of github.com:rest-for-physics/connectorslib into alphaAna

juanangp committed:
Return null pointer when number of signals is zero

juanangp committed:
Merge branch 'master' into alphaAna

juanangp committed:
Merge branch 'master' of github.com:rest-for-physics/connectorslib into
alphaAna

jgalan committed:
Updating GitHub PR badges and template

jgalan reviewed (approved):
I went through the changes. It seems everything fine to me.

jgalan commented:
I just pushed a commit to update the PR badges and template as I already did in other libraries.

juanangp committed:
Merge branch 'master' of github.com:rest-for-physics/connectorslib into alphaAna

juanangp committed:
Addressing validation error

juanangp committed:
Increasing library version
